### PR TITLE
Split BagsExtractor into Extractor and BagsExtractor itself

### PR DESCRIPTION
**Pull request by zurk**
There are already several cases and definitely will be more when we need to process UAST and get some dataset, but as output do not have a BOW model. But we have this fancy pipeline like:
```python
start_point \
        .link(UastRow2Document()) \
        .link(UastDeserializer()) \
        .link(Uast2BagFeatures([extractor]))
```
And it is reasonable to have something similar in case of not-bag data. Before this PR you can actually use the `Uast2BagFeatures` transformer for it. However, some ad-hoc code should be applied to make your `BagsExtracor` work in this case. For example, ~Bag~Extractor to collect Role-Identifiers dataset. 

So I decided to split `BagsExtracor` into `Extractor` and `BagsExtracor` itself.
 
## Changes list
1. `BagsExtractor` splits into `Extractor` and `BagsExtracor`. `BagsExtracor` are mostly the same. `Extractor` is a general wrapper for any UAST algorithm we want to use. As an example see `3.`
2. `Uast2BagFeatures` splited to `Uast2Features` and `Uast2BagFeatures` itself. The first one because now it works with `BagsExtracor`. `Uast2BagFeatures` now it is super simple but produces the same result as before. 
3. `RolesAndIdsExtractor` was completely removed because it was ad-hoc code. In the next PRs, I plan to rewrite other ad-hoc `Extractors`.

This is the main changes to our codebase. As soon as this PR merged, I refactor rest code to use new classes where it is necessary. Also, `BagsExtractor` will use `algorithm` field. now it is omitted on purpose

### CODE REVIEW

File: sourced/ml/extractors/bags_extractor.py [hunk 1]
```diff
@@ -18,7 +51,7 @@ def __init__(self, docfreq_threshold=None, weight=None, **kwargs):
         :param weight: TF-IDF will be multiplied by this weight to change importance of specific \
                       bag extractor
         """
-        super().__init__(**kwargs)
+        super().__init__(algorithm=None, **kwargs)
```
Comment (vmarkovtsev) [inline sourced/ml/extractors/bags_extractor.py]: thus this should go away

File: sourced/ml/extractors/bags_extractor.py [hunk 2]
```diff
@@ -1,14 +1,47 @@
...
+    Converts a single UAST via `algorithm` to anything you need.
+    It is a wrapper to use in `Uast2Features` Transformer in a pipeline.
+    """
+    NAME = None  # feature scheme name, should be overridden in the derived class.
+    OPTS = dict()  # cmdline args which are passed into __init__()
+
+    def __init__(self, algorithm: Callable, name: Optional[str]=None, **kwargs):
+        super().__init__(**kwargs)
+        if name is not None:
+            self.NAME = name
```
[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply sourced/ml/extractors/bags_extractor.py]: You must rely on the abstract `extract()` and remove the constructor.

File: sourced/ml/transformers/uast2bag_features.py [hunk 3]
```diff
@@ -28,7 +28,27 @@ def documentize(self, r: Row) -> Row:
         return Row(**{bfc.document: doc, ec.Uast: r[ec.Uast]})
 
 
-class Uast2BagFeatures(Transformer):
+class Uast2Features(Transformer):
+    def __init__(self, extractors: Union[Extractor, Iterable[Extractor]], **kwargs):
```
[the remark this answers is not in the document]
Comment (zurk) [reply sourced/ml/transformers/uast2bag_features.py]: good idea!

File: sourced/ml/extractors/bags_extractor.py [hunk 4]
```diff
@@ -1,14 +1,44 @@
...
+class Extractor(PickleableLogger):
+    """
+    Converts a single UAST via `algorithm` to anything you need.
+    It is a wrapper to use in `Uast2Features` Transformer in a pipeline.
+    """
+    NAME = None  # feature scheme name, should be overridden in the derived class.
+    ALGORITHM = None  # algorithm class to extract from UAST
+    OPTS = dict()  # cmdline args which are passed into __init__()
+
+    def __init__(self, **kwargs):
```
Comment (vmarkovtsev) [inline sourced/ml/extractors/bags_extractor.py]: You can remove this empty __init__

File: sourced/ml/extractors/bags_extractor.py [hunk 5]
```diff
@@ -1,14 +1,44 @@
...
+    def get_kwargs_fromcmdline(cls, args):
+        prefix = cls.NAME + "_"
+        result = {}
+        for k, v in args.__dict__.items():
+            if k.startswith(prefix):
+                result[k[len(prefix):]] = v
+        return result
+
+    def extract(self, uast: bblfsh.Node):
+        yield from self.ALGORITHM(uast)
```
Comment (vmarkovtsev) [inline sourced/ml/extractors/bags_extractor.py]: Good :+1:

File: sourced/ml/extractors/bags_extractor.py (hunk 1 above)
Comment (vmarkovtsev) [reply sourced/ml/extractors/bags_extractor.py]: All right, this must be updated for `ALGORITHM`

File: sourced/ml/transformers/uast2bag_features.py [hunk 6]
```diff
@@ -28,26 +28,34 @@ def documentize(self, r: Row) -> Row:
...
-    def __init__(self, extractors: Iterable[BagsExtractor], **kwargs):
+class Uast2Features(Transformer):
+    def __init__(self, *extractors: Extractor, **kwargs):
         super().__init__(**kwargs)
         self.extractors = extractors
 
     def __call__(self, rows: RDD):
         return rows.flatMap(self.process_row)
 
+    def process_feature(self, row: Row, name, feature):
```
Comment (vmarkovtsev) [inline sourced/ml/transformers/uast2bag_features.py]: This must be implemented in the derived class.

UastMiner -> Uast2FeaturesDict, Uast2BagOfFeatures

(I am open to better names)

Comment (zurk) [reply sourced/ml/transformers/uast2bag_features.py]: UastMiner, Uast2Features, Uast2BagFeatures?

**Pull request merged**